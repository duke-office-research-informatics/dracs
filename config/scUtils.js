import React from "react";
import { ThemeProvider } from "styled-components";
import { mount, shallow } from "enzyme";
import { theme as defaultTheme } from "../lib/dracs.es.js";

export const shallowWithTheme = (children, customTheme) => {
  const theme = customTheme || defaultTheme;
  return shallow(children, { context: theme });
};

/* eslint-disable */
export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
    ),
  });
}

//FOR ACCESSING PORTALS VIA ENZYME -- import portal from dracs
// https://github.com/airbnb/enzyme/issues/1202#issuecomment-351639783
// function mountPortal(el){
//   const wrapper =  mountWithTheme(el);
//   wrapper.setState({ visible: true});
//   const portalInstance = wrapper.find(Portal).instance();
//   return mountWithTheme(portalInstance.props.children);
// }

//FOR ACCESSING EVENT LISTENERS ATTACHED TO DOCUMENT OR DOCUMENT.BODY
//https://github.com/airbnb/enzyme/issues/426#issuecomment-296304015
// const map = {};
// document.body.addEventListener = jest.fn((event, cb) => {
//   map[event] = cb;
// });
