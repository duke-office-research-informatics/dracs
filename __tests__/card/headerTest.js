import React from "react";
import { mountWithTheme } from "../../config/scUtils.js";
import { axe } from "../../config/axe-helper.js";
import { theme, CardHeader, IconMenu } from "../../lib/dracs.es.js";

describe("CardHeader", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithTheme(<CardHeader />);
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("returns no a11y violations", async () => {
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  it("sets a bottom border when `border` is set to true", () => {
    wrapper.setProps({ border: true });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("header__CardHeaderWrap")).toHaveStyleRule(
      "border-bottom",
      `solid 1px ${theme.colors.border}`
    );
  });

  it("renders children", () => {
    const header = mountWithTheme(
      <CardHeader>
        <IconMenu />
      </CardHeader>
    );
    expect(header).toMatchSnapshot();
    expect(header.find(IconMenu)).toBeDefined();
  });

  it("passes a classname via props", () => {
    wrapper.setProps({ className: "testClass" });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("header__CardHeaderWrap").hasClass("testClass")).toBe(
      true
    );
  });

  it("applies the proper styles for drag-n-drop when `dragHandle` is true", () => {
    wrapper.setProps({ dragHandle: true });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("header__CardHeaderWrap")).toHaveStyleRule(
      "cursor",
      "move"
    );
    expect(
      wrapper.find("header__CardHeaderWrap")
    ).toHaveStyleRule("background-color", "rgba(0,0,0,0.1)", {
      modifier: ":hover",
    });
  });

  it("sets an html title via props", () => {
    const title = '<div id="title">title</div>';
    wrapper.setProps({ htmlTitle: true, title: title });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("#title")).toBeDefined();
  });

  it("sets the icon via props", () => {
    wrapper.setProps({ icon: <IconMenu /> });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(IconMenu)).toBeDefined();
  });

  it("attaches a style object to the header", () => {
    wrapper.setProps({ style: { backgroundColor: "#fff" } });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("header__CardHeaderWrap").props().style).toHaveProperty(
      "backgroundColor",
      "#fff"
    );
  });
});
