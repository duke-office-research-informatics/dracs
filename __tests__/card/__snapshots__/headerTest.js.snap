// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardHeader applies the proper styles for drag-n-drop when \`dragHandle\` is true 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  cursor: move;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c0:hover {
  background-color: rgba(0,0,0,0.1);
}

<header
  className="c0"
/>
`;

exports[`CardHeader attaches a style object to the header 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0"
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
/>
`;

exports[`CardHeader matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0"
/>
`;

exports[`CardHeader passes a classname via props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0 testClass"
/>
`;

exports[`CardHeader renders children 1`] = `
.c1 {
  fill: #fff;
  vertical-align: top;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0"
>
  <svg
    aria-label="open customize and profile menu"
    className="c1"
    color="#fff"
    height={24}
    role="img"
    viewBox="0 0 24 24"
    width={24}
  >
    <title>
      menu icon
    </title>
    <desc>
      hamburger menu icon
    </desc>
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</header>
`;

exports[`CardHeader sets a bottom border when \`border\` is set to true 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  border-bottom: solid 1px #e8e5e2;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0"
/>
`;

exports[`CardHeader sets an html title via props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c1 {
  font-family: "Roboto","Open-Sans",Sans-Serif;
  font-size: 1.5em;
  line-height: 1.35em;
  color: #131516;
  font-weight: 700;
  margin: 4px 0;
  padding: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  margin-right: auto;
  margin-left: 8px;
}

<header
  className="c0"
>
  <h3
    className="c1"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div id=\\"title\\">title</div>",
      }
    }
  />
</header>
`;

exports[`CardHeader sets the icon via props 1`] = `
.c1 {
  fill: #fff;
  vertical-align: top;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 100%;
  padding: 8px;
  overflow: hidden;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<header
  className="c0"
>
  <svg
    aria-label="open customize and profile menu"
    className="c1"
    color="#fff"
    height={24}
    role="img"
    viewBox="0 0 24 24"
    width={24}
  >
    <title>
      menu icon
    </title>
    <desc>
      hamburger menu icon
    </desc>
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</header>
`;
